File: Robotics Kinematics and Dynamics Course/Chapter14/Lesson3/Chapter14_Lesson3.m
function dx = flex_joint_dynamics(t, x, p)
% x = [q; qdot; th; thdot]

q     = x(1);
qdot  = x(2);
th    = x(3);
thdot = x(4);

spring = p.k * (th - q);
damper = p.d * (thdot - qdot);
tau_m  = p.tau0;  % step input

qddot  = (spring + damper - p.m * p.g * p.ell * sin(q) - p.b_l * qdot) / p.J_l;
thddot = (tau_m - spring - damper - p.b_m * thdot) / p.J_m;

dx = [qdot; qddot; thdot; thddot];
end

% Script to simulate
p.J_l = 0.5;  p.J_m = 0.05;
p.b_l = 0.02; p.b_m = 0.01;
p.k   = 150;  p.d   = 0.5;
p.m   = 2.0;  p.ell = 0.4; p.g = 9.81;
p.tau0 = 1.0;

x0 = [0; 0; 0; 0];
[t, x] = ode45(@(t,x) flex_joint_dynamics(t,x,p), [0 2], x0);

q  = x(:,1);
th = x(:,3);
plot(t, q, t, th, '--'); legend('link q', 'motor theta');
xlabel('time [s]'); ylabel('angle [rad]'); grid on;
      

File: Robotics_Kinematics_and_Dynamics_Course/Chapter14/Lesson3/Chapter14_Lesson3.cpp
#include <iostream>
#include <vector>
#include <cmath>

struct FlexParams {
    double J_l{0.5}, J_m{0.05};
    double b_l{0.02}, b_m{0.01};
    double k{150.0}, d{0.5};
    double m{2.0}, ell{0.4}, g{9.81};
    double tau0{1.0};
};

void flex_joint_rhs(const FlexParams& p,
                    double t,
                    const std::vector<double>& x,
                    std::vector<double>& xdot)
{
    double q     = x[0];
    double qdot  = x[1];
    double th    = x[2];
    double thdot = x[3];

    double spring = p.k * (th - q);
    double damper = p.d * (thdot - qdot);
    double tau_m  = p.tau0; // step input

    double qddot  = (spring + damper - p.m * p.g * p.ell * std::sin(q) - p.b_l * qdot) / p.J_l;
    double thddot = (tau_m - spring - damper - p.b_m * thdot) / p.J_m;

    xdot[0] = qdot;
    xdot[1] = qddot;
    xdot[2] = thdot;
    xdot[3] = thddot;
}

int main() {
    FlexParams p;
    double t  = 0.0;
    double dt = 0.0005;
    double T  = 2.0;

    std::vector<double> x(4, 0.0);    // state
    std::vector<double> xdot(4, 0.0); // derivative

    while (t <= T) {
        flex_joint_rhs(p, t, x, xdot);
        // simple explicit Euler integration
        for (int i = 0; i < 4; ++i) {
            x[i] += dt * xdot[i];
        }
        t += dt;
        std::cout << t << " " << x[0] << " " << x[2] << std::endl;
    }
    return 0;
}
      
